### PR TITLE
Enabling sound effects

### PR DESCRIPTION
**Pull request by mkristofik**
This doesn't do anything substantially different from the original code.  But perhaps compiling for a modern Windows allows it work.

### CODE REVIEW

**Comment by jkoppel** [comment]
It's quite odd that I'm not getting E-mails about these PRs when I am for the others.

Quite likely reason: you're sending them off a fork of this repo, instead of off this repo itself. I've added you as a collaborator. Hopefully that will fix things.

File: src/cpp/shared/sound/cd_music.cpp [hunk 1]
```diff
@@ -100,9 +102,9 @@ void bass_play_track(int trknum) {
     return; //this track is already being played
 
   if (!useOpera && trknum >= 4 && trknum <= 9)
-    sprintf(trkbuf, "%s/sw/homm2_%02d.ogg", trkpath, trknum);
+    snprintf(trkbuf, sizeof(trkbuf), "%s/sw/homm2_%02d.ogg", trkpath, trknum);
```
Comment (jkoppel) [inline src/cpp/shared/sound/cd_music.cpp]: Thanks for the small patch. As a small FYI: the reason why I haven't been super-strict about making people use snprintf over sprintf is that it's a bit of a lost cause; there's hundreds of uses throughout the code, usually with fixed-size buffers.

**Comment by jkoppel** [comment]
We'll need to sit down together and figure out why you were having the issue so we can understand how this fixes it

Also, "gbNoSound" is very misnamed. (This name was taken from the original game's code; we didn't name it.) It actually turns off a lot of things. I don't remember why exactly it was necessary to temporarily turn it on (it's been over 5 years, and I was quite a different person then), but that variable does interact with a lot of different subsystems. So, it's not to be messed with lightly.

Similarly, the "PollSound()" function is actually a poor-man's version of non-preemptive concurrency; it should really be thought of as yielding to another thread, which does basic sound/graphics updates.

**Comment by mkristofik** [comment]
`gbNoSound` is controlled by a command-line switch.  I noticed there are numerous if-statements in the disassembly that start with `if (!gbNoSound)` so I set it to zero trying to get my sound to work.  I'll put the original code back the way it was in `init.cpp`.

As for my sound issue itself, I've only been able to determine that it's something to do with the `WAVE_init_driver` function.  In the debugger I saw that the `hdidriver` variable was null and that no sound sample objects were allocated.  My code reproduces what the initialization code did, and somehow that allows it to work.  I'm guessing this is Windows 10 specific, or possibly even specific to my machine.  My last change to exit early should allow this to continue working for people whose sound is already working.

File: src/cpp/shared/sound/cd_music.cpp [hunk 2]
```diff
@@ -222,4 +224,46 @@ void soundManager::CDPlay(int track_number, signed int a3, int a4, int a5) {
...
+  fmt.nAvgBytesPerSec = fmt.nSamplesPerSec;
+  fmt.nBlockAlign = 1;
+  fmt.wBitsPerSample = 8;
+  fmt.cbSize = 0;
+  HDIGDRIVER driver = 0;
+  const int status = AIL_waveOutOpen(&driver, 0, 0, &fmt);
+  if (status == 0) {
+    hdidriver = driver;
+    AllocateSampleHandles();
+  }
```
Comment (jkoppel) [inline src/cpp/shared/sound/cd_music.cpp]: Change this to "} else {"

**Comment by jkoppel** [comment]
The changes you made limit the downside risk, so I can't see a reason not to include this. Make the small formatting change, and then you can merge it yourself.

My guess is that the difference is related to the linker. The way to debug this would be to step through both at the assembly level and look for the difference, which I'll be happy to teach you how to do if you're so interested.

**Pull request merged**